Some general improvements

DarkWiiPlayer opened the pull request.

DarkWiiPlayer committed:
Fix mistake in util.table.map

DarkWiiPlayer committed:
Improve util.deco.if_condition

- The construct `a and b or c` fails for the case `a=true; b=false; c="some value"`, as it returns `c` even though `a` is false
- Since the condition is not a function, its truthyness is known when `if_condition` is called, so a more optimized function can be created at that point.
- Closing functions takes time, so this can be moved outside the `if_condition` function, which also makes intent clearer by giving a name to both functions (`return_first` and `return_second`).

DarkWiiPlayer committed:
Add ldoc comments to util.lua

DarkWiiPlayer committed:
Add type-check to is_instance_of

Desvelao commented on lualog/util.lua:
```diff
@@ -12,36 +16,67 @@ function util.table.includes(t,element)
 function util.is_instance_of(instance,class)
-    return instance.__class_name == class
+    if not type(instance)=='table' then error("Argument #1 not an object (expected table, got "..type(instance)..")", 2) end
+    return and instance.__class_name == class
```
Remove `and` after return statement. Raise an error this line.

Desvelao reviewed (changes requested):
Fix typos (if you can) in function docs:
- `@tparam` to `@param`
- `@treturn` to `@return`

Thank you for your contribution!

DarkWiiPlayer committed:
Remove incorrect "and" keyword

DarkWiiPlayer commented:
> Fix typos (if you can) in function docs

`@treturn` and `@tparam` are special [LDoc tags][ldoc] for typed parameters and return values.

Unless you want compatibility with [LuaDoc (Which doesn't know these tags)][luadoc-tags], they are preferable in this case because they specify the the type of the parameter / return value in addition to its description.

[ldoc]: https://stevedonovan.github.io/ldoc/manual/doc.md.html "LDoc"
[luadoc-tags]: http://keplerproject.github.io/luadoc/manual.html#tags "LuaDoc Tags"

Desvelao commented:
> > Fix typos (if you can) in function docs
> 
> `@treturn` and `@tparam` are special [LDoc tags](https://stevedonovan.github.io/ldoc/manual/doc.md.html) for typed parameters and return values.
> 
> Unless you want compatibility with [LuaDoc (Which doesn't know these tags)](http://keplerproject.github.io/luadoc/manual.html#tags), they are preferable in this case because they specify the the type of the parameter / return value in addition to its description.

Ok, it's perfect then.

DarkWiiPlayer committed:
Add documentation to inspect.lua

DarkWiiPlayer committed:
Simplify allow_tostring option

`tostring()` already checks the `__tostring` metamethod, so there's no need to overcomplicate things.

DarkWiiPlayer commented:
I'm actually starting to wonder how much sense it really makes to have the `Inspect` class when all it really does is one thing. Wouldn't it just be simpler to have a plain function instead?

Desvelao commented:
> I'm actually starting to wonder how much sense it really makes to have the `Inspect` class when all it really does is one thing. Wouldn't it just be simpler to have a plain function instead?

Yes, i guess it would be better. Maybe remove `inpect.lua` file and move the function to `init.lua` as Lualog:inspect(t, options). It would be interesting return self of function to can chain methods.

```lua
function Lualog:inspect(t,options)
   print( inspect_table_function(t, options) ) -- options need to be configured to accept one-time options or self inspect options 
   return self -- to chain methods i mean
end
```

I don't know if it would be interesting accept vararg tables to inspect.

Feel free to try your solution.

DarkWiiPlayer commented:
I think for now I'm done with the "general" improvements. I may have a go at turning inspect into a single function, though it's not like there's anything wrong with the object-oriented approach either.

Feel free to review and merge ♥

Desvelao commented:
Perfect! Thank you! ;)
